domain/domain.go

```diff
@@ -254,121 +254,125 @@ func (do *Domain) GetScope(status string) variable.ScopeFlag {
-			return nil
-		})
-		if err != nil {
-			log.Errorf("mock reload failed err:%v", err)
-			return errors.Trace(err)
-		}
-		return errors.New("mock reload failed")
+		return do.mockReloadFailed()
 	}
 
-	// lock here for only once at same time.
+	// Lock here for only once at the same time.
 	do.m.Lock()
 	defer do.m.Unlock()
 
-	timeout := do.ddl.GetLease() / 2
-	if timeout < defaultMinReloadTimeout {
-		timeout = defaultMinReloadTimeout
+	var err error
+	for {
+		startTime := time.Now()
+		var ver kv.Version
+		ver, err = do.store.CurrentVersion()
+		if err == nil {
+			schemaVersion := int64(0)
+			oldInfoSchema := do.infoHandle.Get()
+			if oldInfoSchema != nil {
+				schemaVersion = oldInfoSchema.SchemaMetaVersion()
+			}
+			err = do.loadInfoSchema(do.infoHandle, schemaVersion, ver.Ver)
+		}
+		if err == nil {
+			atomic.StoreInt64(&do.lastLeaseTS, time.Now().UnixNano())
+			do.SchemaValidity.updateTimeInfo(startTime.UnixNano(), ver.Ver)
+			break
+		}
+		log.Errorf("[ddl] load schema err %v, ver:%v, retry again", errors.ErrorStack(err), ver.Ver)
+		// TODO: Use a backoff algorithm.
+		time.Sleep(doReloadSleepTime)
+		continue
 	}
 
-	exit := int32(0)
-	done := make(chan error, 1)
-	go func() {
-		var err error
-		for {
-			var ver kv.Version
-			ver, err = do.store.CurrentVersion()
-			if err == nil {
-				schemaVersion := int64(0)
-				oldInfoSchema := do.infoHandle.Get()
-				if oldInfoSchema != nil {
-					schemaVersion = oldInfoSchema.SchemaMetaVersion()
-				}
-				err = do.loadInfoSchema(do.infoHandle, schemaVersion, ver.Ver)
+	return errors.Trace(err)
+}
+
+func (do *Domain) checkValidityInLoop(lease time.Duration) {
+	timer := time.NewTimer(lease)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-timer.C:
+			// TODO: Using the local time, it will affect the accuracy of the check when clock transition.
+			lastReloadTime, lastSuccTS := do.SchemaValidity.getTimeInfo()
+			sub := time.Duration(time.Now().UnixNano() - lastReloadTime)
+			if sub > lease {
+				// If sub is greater than a lease,
+				// it means that the schema version hasn't update for a lease.
+				do.SchemaValidity.SetValidity(false, lastSuccTS)
+			} else {
+				do.SchemaValidity.SetValidity(true, lastSuccTS)
 			}
-			if err == nil {
-				atomic.StoreInt64(&do.lastLeaseTS, time.Now().UnixNano())
-				break
+
+			waitTime := lease
+			if sub > 0 {
+				// If the schema is invalid (sub >= lease), it means reload schema will become frequent.
+				// We need to reduce wait time to check the validity more frequently.
+				if sub >= lease {
+					waitTime = lease / 4
+				} else {
+					waitTime -= sub
+				}
 			}
-			do.SchemaValidity.setLastFailedTS(ver.Ver)
-			log.Errorf("[ddl] load schema err %v, retry again", errors.ErrorStack(err))
-			if atomic.LoadInt32(&exit) == 1 {
-				return
+			log.Infof("[ddl] check validity in a loop, sub:%v, lease:%v, succ:%v, waitTime:%v",
+				sub, lease, lastSuccTS, waitTime)
+			timer.Reset(waitTime)
```

Comment:
This is not ideal usage of `timer.Reset`

see https://godoc.org/time#Timer.Reset
> To reuse an active timer, always call its Stop method first and—if it had expired—drain the value from its channel.